File: 74. Search a 2D Matrix.java
/*
Problem Link: https://leetcode.com/problems/search-a-2d-matrix/
*/

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        for (int i = 0; i < matrix.length; i++) {
            if (matrix[i][0] <= target) {
                int l = 0;
                int h = matrix[i].length - 1;
                while (l <= h) {
                    int mid = l + (h - l) / 2;
                    if (matrix[i][mid] == target) {
                        return true;
                    }
                    if (matrix[i][mid] >= target) {
                        h = mid - 1;
                    } else {
                        l = mid + 1;
                    }
                }
            }
        }
        return false;
    }
}

File: Problems/74. Search a 2D Matrix.py
"""
Problem Link: https://leetcode.com/problems/search-a-2d-matrix/
"""


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])

        for i in range(n):
            if matrix[i][0] <= target <= matrix[i][m-1]:
                left = 0
                right = m-1
                while left <= right:
                    mid = (left+right)//2
                    if matrix[i][mid] == target:
                        return True
                    elif matrix[i][mid] > target:
                        right = mid-1
                    else:
                        left = mid+1
        return False
